[Proposal]: Extending Compartment Report (#5)
openbraininstitute/sonata-extension: source/sonata_report.rst

    *   Previously supported values for ``sections`` include ``"soma"``, ``"axon"``, ``"dend"``, ``"apic"``, or ``"all"``. These continue to function as before, typically used with the ``cells`` key to specify target populations.
*   **``compartments``**:
    *   If ``sections`` is ``"compartment_set"``, this field **must** contain the name of a key (a specific compartment set) defined in your ``compartment_sets.json`` file (e.g., ``"example_compartment_set"``).
    *   For other ``sections`` types (``"soma"``, ``"axon"``, ``"dend"``, ``"apic"``, or ``"all"``), ``compartments`` typically takes values like ``"center"`` or ``"all"``.
*   **``cells``**: The ``cells`` key (e.g., ``"cells": "Mosaic"`` or ``"cells": ["popA", 123]``) is **not allowed** and should cause an error if ``sections`` is ``"compartment_set"``. The selection of cells and their specific compartments is entirely managed by the chosen compartment set from ``compartment_sets.json``.

The output HDF5 report format for these targeted compartment reports remains the same as described in the main :ref:`compartment_report_main` section.

> The ordering of axon/apic/dend/soma and the gid, type pair in report can only show up once.

Reviewer: to me this is not clear. `[0, 'dend', ..., ...]` can only appear once? The ordering can only appear once?

Author: > Also add 2 examples, what is not possible and what is possible

Added one example with acceptable entries in `compartment_set` as mentioned in the last comment. Also, added another one for raising a warning for duplicate entries.

        }
    }

Extension of Compartment report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To enable fine-grained control over which specific compartments are recorded, SONATA supports an extension using a ``compartment_sets.json`` file. This allows users to define sets of explicit compartment targets (e.g., ``dend[10](0.1)``) for selected node_id.

File: ``compartment_sets.json``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This JSON file defines named sets of compartment targets. Each set specifies the population and a list of `(node_id, section_name, location)` tuples.

.. code-block:: json

  {
    "example_compartment_set": {
      "population": "S1nonbarrel_neurons",
      "compartment_set": [

>         [0, "dend[10]", 0.1],

Author: Bringing everyone up to speed from the discussion in the issue and comments [here](https://github.com/openbraininstitute/sonata-extension/issues/4#issuecomment-2904447979):  
1. @mgeplf's suggestion is to have 
structure like this `[1, "dend", [[10, 0.2]]]` i.e. `[node_id, section_type, [[section_number, compartment_location]]]`
2. Another suggestion when recording from multiple compartments of the same section is to use `[0, "dend", [[3, 0.1], [5, 0.5], [9, 0.5]]],`

Author: @mgeplf Based on the above comments, should we keep the original structure `[0, "dend[10]", 0.1],`? 

And if it is fine with you, can you approve this PR?

        }
    }

Extension of Compartment report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To enable fine-grained control over which specific compartments are recorded, SONATA supports an extension using a ``compartment_sets.json`` file. This allows users to define sets of explicit compartment targets (e.g., ``dend[10](0.1)``) for selected node_id.

File: ``compartment_sets.json``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This JSON file defines named sets of compartment targets. Each set specifies the population and a list of `(node_id, section_name, location)` tuples.

Reviewer 2: this needs to be updated to reflect the splitting of `type` and `section_index`

.. code-block:: json

  {
    "example_compartment_set": {
      "population": "S1nonbarrel_neurons",
      "compartment_set": [
        [0, "dend", 10, 0.1],
        [1, "dend", 4, 0.1],
        [2, "dend", 3, 0.1],
        [3, "dend", 6, 0.3]
      ]
    }
  }

*   ``population``: The name of the node population these targets belong to.
*   ``compartment_set``: A list of list, where each list is ``[node_id, section_name, section_index, location]``.
    *   ``node_id``: The ID of the node within the specified ``population``.

>     *   ``section_name``: The name of the  NEURON section (e.g., "soma", "dend", "apic"). This should not contain the ``section_index`` e.g. ``"dend[10]"``

Reviewer 2: should we enumerate all of them? ie: add `axon`? and not allow anything that isn't enumerated?

Author: @cattabiani We cannot use custom section names with neurodamus, e.g. `AIS` and `Hilock`. If we can't right now, is it possible to allow it in future?

Reviewer: Endfeet?

Reviewer: Ah no. I confused the location of this comment. Forget the endfeet

    *   ``section_index``: The number of the NEURON section index e.g. for dend[10], ``section_index`` is 10. Smallest ``section_index`` is 0 in NEURON.

>     *   ``location``: The fractional distance along the section (0.0 to 1.0).

Reviewer 2: are `(` and `)` being used in the interval sense?  ie: is it `0<= location <= 1` or `0 < location < 1` or something else?

> This file is typically placed alongside other simulation configuration files, for example:

Reviewer 2: I don't think this needs to be included; I would say that the file is subject to the `manifest` described in the sonata spec: "The "manifest" section of the config file provides a convenient handle on setting variables that point to base paths."; people can organize their files how they want within those constraints.

Author: Removed the file location text paragraph.

    }
    // ... other global configurations ...
  }

Key changes:

*   **``sections``**:
    *   When set to ``"compartment_set"``, it indicates that the report targets are defined by a named set in the ``compartment_sets.json`` file.
    *   Previously supported values for ``sections`` include ``"soma"``, ``"axon"``, ``"dend"``, ``"apic"``, or ``"all"``. These continue to function as before, typically used with the ``cells`` key to specify target populations.
*   **``compartments``**:
    *   If ``sections`` is ``"compartment_set"``, this field **must** contain the name of a key (a specific compartment set) defined in your ``compartment_sets.json`` file (e.g., ``"example_compartment_set"``).
    *   For other ``sections`` types (``"soma"``, ``"axon"``, ``"dend"``, ``"apic"``, or ``"all"``), ``compartments`` typically takes values like ``"center"`` or ``"all"``.
*   **``cells``**: The ``cells`` key (e.g., ``"cells": "Mosaic"`` or ``"cells": ["popA", 123]``) is **not allowed** and should cause an error if ``sections`` is ``"compartment_set"``. The selection of cells and their specific compartments is entirely managed by the chosen compartment set from ``compartment_sets.json``.

> The output HDF5 report format for these targeted compartment reports remains the same as described in the main :ref:`compartment_report_main` section.

Reviewer 2: The *ordering* needs to be clearly specified; one has to be able to go from the list of compartments to a total ordering in the report file, programatically, without ambiguity.

ie: what happens if there is the same gid multple times? how are the orders concatenated (this was partly why I asked about doing `[gid, type, [[section, location], ...]` so there is no ambiguity; just say that there is an ordering of `axon/apic/dend/soma` and the `gid, type` pair can only show up once.

Reviewer 2: > Is it enough for the compartment_set ordering just be one-to-one with the report ordering?

It has to conform to the way reports are laid out:
https://sonata-extension.readthedocs.io/en/latest/sonata_report.html#compartment-report
<img width="706" alt="image" src="https://github.com/user-attachments/assets/f8352e81-1c53-4892-9e20-6782c9dd65cb" />

Author: Just for clarity, there are two different orderings we are talking about here:
1. Order of entries in the compartment_set.json
2. Order of output entries in the h5 file.

Point 1:
The order of entries can be any order of gids
`[1, "dend", 10, 0.1],`
`[2, "dend", 10, 0.7],`
`[0, "dend", 10, 0.4],`
`[1, "dend", 10, 0.3],`

Point 2:
For what you mentioned here @mgeplf :
> Ok, I guess having ordering isn't important; I thought that it was needed to more easily tie together the compartment_set that generated a particular report value, but that's probably not the most import thing.

We need to follow the same orders in the h5 file as the entries in the compartment_set.json in order to read and perform analysis on those entries. Or have I misunderstood anything here.

Reviewer: obviously you can remove gids. the order remains the same

Author: > I'm wondering why the location (section_type, section_index_within_type, pos) was chosen - SONATA generally > encodes just (absolute_section_id, pos): both in the reports, and in the synapse locations. It seems like an 
> unnecessary deviation from the standard - is it required?

If you meant the list here `[0, "dend", 10, 0.1]`, this was done for ease of understanding and the way we will create `compartment_set` entries in obi-one. 

>  obviously you can filter by gid preserving relative order of the remaining entries

Alessandro says above that we can easily filter with gids. So, I am assuming the reporting should be fine.

Reviewer 2: > If you meant the list here [0, "dend", 10, 0.1], this was done for ease of understanding and the way we will create compartment_set entries in obi-one.

Yes, that is what I meant.
How hard would it be to translate to the "normal SONATA" form for obi-wan @james-isbister?
Having different addressing schemes has bitten us in the past (and we still have technical debt from that), so I'd prefer to not create a *third* way of addressing if possible.

Reviewer 2: (by the way, sorry about only noticing this now, I should have caught it before)

Reviewer: I see that just below there is `section_id` and `offset`. IIUC `section_id = (section_name, section_index)` and `offset = location`. Should we use these names instead? Maybe keeping section_id split is better (imho at least) but location is offset. We should coalesce and always use 1 term for a certain thing no? Are we Tolkien's elves? (they like to rename stuff over and over)

Reviewer 3: If I understand correctly the suggestion is that we get rid of "dend" from  [0, "dend", 10, 0.1] because its not needed, and we instead use tuples of the form (gid, absolute_section_id, pos) in the compartment set json file? Moreover, these tuples will be stored in the output report h5 file?

That would definitely work for us, as we haven't implemented it yet in obi-one, and want to go with the best way.

The only question would be, whether we need to to order the tuples/lists (0, 10, 0.1) in the compartment_set json file that we generate? Or ordering is just for the generated output report h5 files?

Reviewer: is `absolute_section_id` a progressive counter that combines `section_name` and `section_index`? Do you know a part of the code in libsonata that uses this?

For me no ordering of the inputs is required:
- if there are duplicates I can emit a warning (without spamming)
- the h5 file will follow the same order minus the gids that are not present in the selection

Let me know if I did not understand something

Author: @mgeplf Does the `[gid, absolute_section_id, pos]` sound like a good way to write `comaprtment_sets` as James suggests above or do you want me to change to something else? We would need to map the gid to each recording. Is it stored in the `mapping` entry in the report?

Reviewer 2: @darshanmandge >  Does the [gid, absolute_section_id, pos] sound like a good way to write comaprtment_sets as James suggests above or do you want me to change to something else?

That sounds reasonable to me.

> We would need to map the gid to each recording. Is it stored in the mapping entry in the report?

My understanding is that the gid is stored in `/report/{population_name}/mapping/node_ids`.

> Edit: How can a user create such entries in compartment_set.json with bluecellulab or neurodamus: [gid, absolute_section_id, pos]? The user generally knows the NEURON compartment location e.g. dend[10](0.3). How can one get the absolute_section_id ?

I think the same way that the entries in the edge files were created.
In that case the whole morphology is loaded, and the calculation is done that way.
In the case of neuron, I think one could count and sum the number of sections in the types that precede the desired type (in the order defined by the SONATA spec https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#swc-format-additional-requirements), and then add the sections from the current type, but rooted before the neurite in question and then the ones that compose the current neurite.

@cattabiani > What do we do if we have multiple reports? In that case a post-processing step where we merge reports is necessary

I think the best way to handle that is to disallow multiple reports having the same `file_name` if they are addressing the same population; but we should add that to the spec - then we don't have to worry about collisions.

@cattabiani > the current implementation has a limitation: you cannot get a report with gids from different populations. Is it ok?

I think it's fine; one can just create multiple reports to include different populations; in this case they could even use the same `file_name`, since the populations are disjoint.

Reviewer: ok, things are getting clearer (to me). I have 2 additional knitpicks about naming. I think it is important to be consistent and have 1 and only 1 term to define one "thing". It is also important to try to respect the naming also in the code so that new developers can quickly get up to speed. Thus:
- given the sonata docs I think the correct term should be `offset` and not `location` for the last [0, 1] double
- we did not specify how to name a list `[gid, section_id, offset]`. I like `CompartmentLocation` but I am very open to suggestions

Reviewer: yes, ask for a review when you are ready

Author: @ all Changed to `[node_id, section_id, offset]` format for `compartment_set` entries in the latest commit.

Reviewer: I just noticed that in the report it is refferred to as element_id. Definetely I would stick close to that. id better than index. If you prefer section let's keep section

Reviewer: >IIRC, that's because there isn't a 1-1 mapping between compartment and sections/offset; it depends on how what nseg is set to.
@mgeplf doesn't `element_ids` refer to sections (not compartments/segments)? I do not fully get this comment. The segment selection is done by the offset double no?

Reviewer 2: I think the exact opposite: we know this is a `section`, so we should name it as such.
The compartment reports use a different name, because the contents of the report aren't necessarily section based: 
![image](https://github.com/user-attachments/assets/69115acc-c18d-4fb5-8bc9-17b264d81442)

        }
    }

Extension of Compartment report
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To enable fine-grained control over which specific compartments are recorded, SONATA supports an extension using a ``compartment_sets.json`` file. This allows users to define sets of explicit compartment targets (e.g., ``dend[10](0.1)``) for selected node_id.

File: ``compartment_sets.json``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This JSON file defines named sets of compartment targets. Each set specifies the population and a list of ``[node_id, section_id, offset]`` lists:

.. code-block:: json

  {
    "example_compartment_set": {
      "population": "S1nonbarrel_neurons",
      "compartment_set": [
        [0, 1, 0.1],
        [3, 5, 0.7],
        [0, 2, 0.3]
      ]
    }
  }

> *   ``population``: The name of the node population of the ``node_id``s in ``compartment_set``.

Reviewer 4: As I checked, without the "\", the rendered output is in the wrong format. 
```suggestion
*   ``population``: The name of the node population of the ``node_id``\s in ``compartment_set``.
```

*   ``compartment_set``: A list of lists, where each list is ``[node_id, section_id, offset]``. Each of the ``[node_id, section_id, offset]`` lists is referred to as a ``CompartmentLocation``.
*   ``node_id``: The ID of the node within the specified ``population``.

> *   ``section_index``: The global index of a given section within its cell. NOTE: this is NOT the NEURON section index (e.g. 10 for ``dend[10]``). The ``section_index`` is calculated similar to ``get_section_index`` function of the neurodamus repository in `neurodamus.reports.py <https://github.com/openbraininstitute/neurodamus/blob/1e8b00e55bcc08e9047d6c9a48d068c463c53aef/neurodamus/report.py#L6>`_.

Reviewer: In neurodamus it is section_index at least in one function. Also in my implementation in libsonata I used this convention. Why do you argue for section_id?

*   ``offset``: The fractional distance along the section (0<= offset <=1). NOTE: offset for a section is similar to the NEURON segment location e.g. 0.5 in ``dend[10](0.5)``.

> The order of ``node_id`` in ``compartment_set`` is not necessary. Such entries are acceptable:

Reviewer: this is ambyguous. What does it mean "not necessary"? Is it sorted automatically or is it kept as is? 
```suggestion
The code preserves the order of ``node_id`` in ``compartment_set``. Such entries are acceptable:
```

.. code-block:: json
    
    {
        "compartment_set": [
            [0, 1, 0.1],
            [2, 5, 0.7],
            [0, 2, 0.3]
        ]
    }

> However, there should be a warning if they are the same, but even complete duplicates can be handled such as

Reviewer: I added the warning in the beginning. Then Mike didn't like it. Also it is not so straight forward because I need to compare doubles. Do I do it with some approximation? Then the actual threshold becomes a hot topic

    "example_compartment_set": {
      "population": "S1nonbarrel_neurons",
      "compartment_set": [
        [0, "dend", 10, 0.1],
        [1, "dend", 4, 0.1],
        [2, "dend", 3, 0.1],
        [3, "dend", 6, 0.3]
      ]
    }
  }

*   ``population``: The name of the node population these targets belong to.
*   ``compartment_set``: A list of list, where each list is ``[node_id, section_name, section_number, location]``.
    *   ``node_id``: The ID of the node within the specified ``population``.
    *   ``section_name``: The name of the section (e.g., "soma", "dend", "apic"). This should not contain the section number e.g. "dend[10]"

>     *   ``section_number``: The number of the section (e.g., "soma", "dend", "apic").

Reviewer 5: Is this section number zero-based? If so, how about calling it `section_index` and mention this in the description as well, just to avoid any confusion?

      "population": "S1nonbarrel_neurons",
      "compartment_set": [
        [0, 1, 0.1],
        [0, 2, 0.3]
        [3, 5, 0.7],
      ]
    }
  }

* ``population``: The name of the node population of the ``node_id``\s in ``compartment_set``.
* ``compartment_set``: A list of lists, where each list is ``[node_id, section_id, offset]``. Each of the ``[node_id, section_id, offset]`` lists is referred to as a ``CompartmentLocation``.
* ``node_id``: The ID of the node within the specified ``population``.
* ``section_id``: The global index of a given section within its cell.
* ``offset``: The fractional distance along the section (0<= offset <=1). NOTE: offset for a section is similar to the NEURON segment location e.g. 0.5 in ``dend[10](0.5)``.

.. note::

>    The ``section_id`` is NOT the NEURON section index (e.g. 10 for ``dend[10]``).

Reviewer: you need to explain what is the `section_id` in this case. you are just saying what it is not

Reviewer 2: it's described a few lines up

Reviewer: yes, I am blind

Key changes:

**sections**:

* When set to ``"compartment_set"``, it indicates that the report targets are defined by a named set in the ``compartment_sets.json`` file.
* Previously supported values for ``sections`` include ``"soma"``, ``"axon"``, ``"dend"``, ``"apic"``, or ``"all"``. These continue to function as before, typically used with the ``cells`` key to specify target populations.

**compartments**:

* If ``sections`` is ``"compartment_set"``, this field **must** contain the name of a key (a specific compartment set) defined in your ``compartment_sets.json`` file (e.g., ``"example_compartment_set"``).
* For other ``sections`` types (``"soma"``, ``"axon"``, ``"dend"``, ``"apic"``, or ``"all"``), ``compartments`` typically takes values like ``"center"`` or ``"all"``.

**cells**:

> * The ``cells`` key (e.g., ``"cells": "Mosaic"`` or ``"cells": ["popA", 123]``) is **not allowed** and should cause an error if ``sections`` is ``"compartment_set"``. The selection of cells and their specific compartments is entirely managed by the chosen compartment set from ``compartment_sets.json``.

Reviewer 2: ok, I added that.
